### PR TITLE
feat(logger): customizing logger output

### PR DESCRIPTION
**Pull request by michaelbrewer**
**Issue #, if available:** #138 and #143

## Description of changes:

> **NOTE:** `sampling_rate` and `service` can not be suppressed

Ensure that the initial key seq is: `level,location,message,timestamp` so by default messages in Cloudwatch logs are easier to read:

```python3
>>> logger = Logger()
>>> logger.info("Eggs")
{'level': 'INFO', 'location': 'method_name:126', 'message': 'Eggs', 'timestamp': '2020-08-27 15:32:51,285', 'service': 'app', 'sampling_rate': 0.0}
```

Also allow for reordering override via `format_keys`

```python3
>>> logger = Logger(format_keys=["message"])
>>> logger.info("Foo")
{'message': 'Message', 'level': 'INFO', 'location': 'method_name:139', 'timestamp': '2020-08-27 15:36:11,494', 'service': 'app', 'sampling_rate': 0.0}
```

You can also suppress `timestamp` or `location` fields by passing them in as `None`

```python3
>>> logger = Logger(stream=stdout, location=None, timestamp=None)
>>> logger.info("magnus")
{"level": "INFO", "message": "magnus", "service": "app", "sampling_rate": 0.0}
```

Finally you can already customize the format for `location` and `timestamp`. So putting it all together:
  
```python3
>>> logger = Logger(location="%(module)s.%(funcName)s:%(lineno)d", timestamp=None, format_keys=["location", "message"])
>>> logger.info("takeoff")
{"location": "space.fly:7", "message": "takeoff", "level": "INFO", "service": "falcon", "sampling_rate": 0.0}
```

**Checklist**

 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)

### CODE REVIEW

**Comment by michaelbrewer** [comment]
@heitorlessa  maybe we can make the order configurable or get some feedback on what it should be.

File: aws_lambda_powertools/logging/formatter.py [hunk 1]
```diff
@@ -44,16 +22,15 @@ def __init__(self, **kwargs):
 
         Other kwargs are used to specify log field format strings.
         """
-        self.default_json_formatter = kwargs.pop("json_default", json_formatter)
+        self.default_json_formatter = kwargs.pop("json_default", str)
```
Comment (heitorlessa) [inline aws_lambda_powertools/logging/formatter.py:25]: add a comment that unserializable values will be cast as str

File: aws_lambda_powertools/logging/logger.py [hunk 2]
```diff
@@ -292,6 +273,29 @@ def structure_logs(self, append: bool = False, **kwargs):
                 handler.setFormatter(JsonFormatter(**self._default_log_keys, **kwargs))
 
 
+def _get_log_level(level: Union[str, int]) -> Union[str, int]:
```
Comment (heitorlessa) [inline aws_lambda_powertools/logging/logger.py:276]: make this a classmethod, not a separate function - it's still part of the encapsulated logger class logic

File: aws_lambda_powertools/logging/logger.py [hunk 3]
```diff
@@ -292,6 +273,29 @@ def structure_logs(self, append: bool = False, **kwargs):
...
+    if isinstance(level, int):
+        return level
+
+    log_level: str = level or os.getenv("LOG_LEVEL")
+    log_level = log_level.upper() if log_level is not None else logging.INFO
+
+    return log_level
+
+
+def _get_caller_filename():
```
Comment (heitorlessa) [inline aws_lambda_powertools/logging/logger.py:287]: make this a classmethod, not a separate function - it's still part of the encapsulated logger class logic

File: aws_lambda_powertools/logging/formatter.py [hunk 4]
```diff
@@ -44,16 +22,15 @@ def __init__(self, **kwargs):
...
         datefmt = kwargs.pop("datefmt", None)
 
         super(JsonFormatter, self).__init__(datefmt=datefmt)
         self.reserved_keys = ["timestamp", "level", "location"]
-        self.format_dict = {
-            "timestamp": "%(asctime)s",
-            "level": "%(levelname)s",
-            "location": "%(funcName)s:%(lineno)d",
-        }
+        self.format_dict = dict.fromkeys(kwargs.pop("format_keys", ["level", "location", "message", "timestamp"]))
```
Comment (heitorlessa) [inline aws_lambda_powertools/logging/formatter.py:30]: could you please add a comment about insertion order here?

**Review by heitorlessa — CHANGES_REQUESTED** [review]
made a few comments ;)

Additional thoughts:

* I'm not 100% sure `format_keys` is descriptive enough for whoever unfamiliar with powertools to read and understand what's going on
* You had a comment about suppressing `sampling_rate` and possibly others in the future, has that changed? This doesn't address that

**Comment by michaelbrewer** [comment]
> made a few comments ;)
> 
> 
> 
> Additional thoughts:
> 
> 
> 
> * I'm not 100% sure `format_keys` is descriptive enough for whoever unfamiliar with powertools to read and understand what's going on
> 

Maybe ordered_message_keys? Kind of unsure of the name. :)


> * You had a comment about suppressing `sampling_rate` and possibly others in the future, has that changed? This doesn't address that

For `service` and `sample_rate` I would have to add a check for if it was set to None, vs just being left out. I will have to think about how.

File: aws_lambda_powertools/logging/formatter.py (hunk 1 above)
Comment (michaelbrewer) [reply aws_lambda_powertools/logging/formatter.py:25]: @heitorlessa The original code did mention this in the method docstring. I can put in a line comment as well.

File: aws_lambda_powertools/logging/logger.py (hunk 3 above)
Comment (michaelbrewer) [reply aws_lambda_powertools/logging/logger.py:287]: @heitorlessa should I use `@staticmethod` then ?

File: aws_lambda_powertools/logging/formatter.py (hunk 1 above)
Comment (heitorlessa) [reply aws_lambda_powertools/logging/formatter.py:25]: a line comment suffices ;)

File: aws_lambda_powertools/logging/logger.py (hunk 3 above)
Comment (heitorlessa) [reply aws_lambda_powertools/logging/logger.py:287]: sorry, yes, that's what I meant - `staticmethod`.

**Comment by heitorlessa** [comment]
> > made a few comments ;)
> > Additional thoughts:
> > 
> > * I'm not 100% sure `format_keys` is descriptive enough for whoever unfamiliar with powertools to read and understand what's going on
> 
> Maybe ordered_message_keys? Kind of unsure of the name. :)
> 
> > * You had a comment about suppressing `sampling_rate` and possibly others in the future, has that changed? This doesn't address that
> 
> For `service` and `sample_rate` I would have to add a check for if it was set to None, vs just being left out. I will have to think about how.

What about `log_record_order`? Seems explicit enough to me even as sleepy as I am now. 

I'd suggest keeping as is in terms of suppression, there could be side effects on these two:

* If the customer wants to use the dynamic sampling feat (I have a list of customers already using in prod)
* `service` is integral to Powertools terminologies, and this will degrade ops capabilities when creating metrics/visualization from logs etc

**Comment by michaelbrewer** [comment]
I will keep the log suppression as it then. We should make have a better overview of Powertools in the main website which highlights this. I still don't quite understand log sampling :-)

File: tests/functional/test_aws_lambda_logging.py [hunk 5]
```diff
@@ -116,3 +116,57 @@ class Unserializable:
...
+    logger.info("Message")
+
+    log_dict: dict = json.loads(stdout.getvalue())
+
+    # THEN the beginning key sequence must be `level,location,message,timestamp`
+    assert ",".join(list(log_dict.keys())[:4]) == "level,location,message,timestamp"
+
+
+def test_log_dict_key_custom_seq(stdout):
+    # GIVEN a logger configuration with format_keys set to ["message"]
```
Comment (heitorlessa) [inline tests/functional/test_aws_lambda_logging.py:135]:
```suggestion
    # GIVEN a logger configuration with log_record_order set to ["message"]
```

**Comment by heitorlessa** [comment]
Thank you @michaelbrewer - I've just noticed I don't have push permissions to your branch to update a tiny typo in the test comments, so I'm merging as-is and will fix-later.

**Comment by michaelbrewer** [comment]
> Thank you @michaelbrewer - I've just noticed I don't have push permissions to your branch to update a tiny typo in the test comments, so I'm merging as-is and will fix-later.

Oops. Thanks.

**Pull request merged**